train.py

```diff
@@ -3,340 +3,388 @@
+    # Seeding can be annoying in pytorch at the moment. Based on my experience, the below means of seeding
+    # allows for deterministic experimentation.
+    torch.manual_seed(args.seed)
+    np.random.seed(args.seed)  # set seed
+    random.seed(args.seed)
+    device = (
+        torch.cuda.current_device()
+        if torch.cuda.is_available() and args.num_gpus_to_use > 0
+        else "cpu"
+    )
+    args.device = device
+    if torch.cuda.is_available():
+        torch.cuda.manual_seed_all(args.seed)
+        torch.backends.cudnn.deterministic = True
+
+    # Always save a snapshot of the current state of the code. I've found this helps immensely if you find that one of your
+    # many experiments was actually quite good but you forgot what you did
+
+    snapshot_filename = "{}/snapshot.tar.gz".format(saved_models_filepath)
+    filetypes_to_include = [".py"]
+    all_files = []
+    for filetype in filetypes_to_include:
+        all_files += glob.glob("**/*.py", recursive=True)
+    with tarfile.open(snapshot_filename, "w:gz") as tar:
+        for file in all_files:
+            tar.add(file)
+
+    ######################################################################################################### Model
+
+    model = get_model(
+        model_type=args.model.type,
+        num_classes=num_classes,
+        dataset_name=args.dataset_name,
+    ).to(device)
+
+    if args.num_gpus_to_use > 1:
+        model = nn.parallel.DistributedDataParallel(
+            model
+        )  # more efficient version of DataParallel
+
+    model = model.to(device)
+
+    ######################################################################################################### Optimisation
+
+    params = model.parameters()
+    criterion = nn.CrossEntropyLoss()
+
+    if args.optim.lower() == "sgd":
+        optimizer = optim.SGD(
+            params,
+            lr=args.learning_rate,
+            momentum=args.momentum,
+            weight_decay=args.weight_decay,
+        )
+    else:
+        optimizer = optim.Adam(
+            params, lr=args.learning_rate, weight_decay=args.weight_decay
+        )
+
+    if args.scheduler == "CosineAnnealing":
+        scheduler = CosineAnnealingLR(
+            optimizer=optimizer, T_max=args.max_epochs, eta_min=0
+        )
+    else:
+        scheduler = MultiStepLR(optimizer, milestones=args.milestones, gamma=0.2)
+
+    ######################################################################################################### Restoring
+
+    restore_fields = {
+        "model": model,
+        "optimizer": optimizer,
+        "scheduler": scheduler,
+    }
+
+    start_epoch = 0
+    if args.resume:
+        resume_epoch = restore_model(restore_fields, path=saved_models_filepath)
+        if resume_epoch == -1:
+            print("Failed to load from {}/ckpt.pth.tar".format(saved_models_filepath))
```

Comment:
If this fails should we not exit the code? Basically if the use wanted to resume and couldn't, the flow of logic should then be "oh well, restart time".

Reply:
We could either exit, or start from scratch. It depends on whether we can tolerate restarts albeit with a message, or whether we'd rather throw an informative error message that can ask the user to remove the resume flag.

Reply:
I'd rather for for the latter tbh.